dojo/utils.py

```diff
@@ -2108,3 +2121,102 @@ def is_scan_file_too_large(scan_file):
         if size > settings.SCAN_FILE_MAX_SIZE:
             return True
     return False
+
+
+def sla_compute_and_notify(*args, **kwargs):
+    """
+    The SLA computation and notification will be disabled if the user opts out
+    of the Findings SLA on the System Settings page.
+
+    Notifications are managed the usual way, so you'd have to opt-in.
+    Exception is for JIRA issues, which would get a comment anyways.
+    """
+    def _notify(finding, title):
+        create_notification(
+            event='sla_breach',
+            title=title,
+            finding=finding,
+            sla_age=sla_age
+        )
+
+        if do_jira_sla_comment:
+            logger.info("Creating JIRA comment to notify of SLA breach information.")
+            add_simple_jira_comment(jira_config, jira_issue, title)
+
+    # exit early on flags
+    if not settings.SLA_NOTIFY_ACTIVE and not settings.SLA_NOTIFY_ACTIVE_VERIFIED_ONLY:
+        logger.info("Will not notify on SLA breach per user configured flags")
+        return
+
+    jira_issue = None
+    jira_config = None
+    try:
+        system_settings = System_Settings.objects.get()
+        if system_settings.enable_finding_sla:
+            logger.info("About to process findings for SLA notifications.")
+            logger.debug("Active {}, Verified {}, Has JIRA {}, pre-breach {}".format(
+                settings.SLA_NOTIFY_ACTIVE,
+                settings.SLA_NOTIFY_ACTIVE_VERIFIED_ONLY,
+                settings.SLA_NOTIFY_WITH_JIRA_ONLY,
+                settings.SLA_NOTIFY_PRE_BREACH,
+            ))
+
+            query = None
+            if settings.SLA_NOTIFY_ACTIVE:
+                query = Q(active=True, is_Mitigated=False, duplicate=False)
+            if settings.SLA_NOTIFY_ACTIVE_VERIFIED_ONLY:
+                query = Q(active=True, verified=True, is_Mitigated=False, duplicate=False)
+            logger.debug("My query: {}".format(query))
+
+            no_jira_findings = {}
+            if settings.SLA_NOTIFY_WITH_JIRA_ONLY:
+                logger.debug("Ignoring findings that are not linked to a JIRA issue")
+                no_jira_findings = Finding.objects.exclude(jira_issue__isnull=False)
+            findings = Finding.objects \
+                .filter(
+                    query
+                ).exclude(id__in=no_jira_findings)
```

Comment:
if you add the filter directly in the query it will be a lot faster I think.

the no_jira_findings query to me looks it will be all findings WITH a jira_issue, so inverted?

Reply:
I think you also want to do some prefetching because otherwise the part below will be very slow and hog your celery if you have 500k findings like I know you have :-)

for example something like this:

```
        prefetched_findings = prefetched_findings.select_related('reporter')
        prefetched_findings = prefetched_findings.select_related('jira_issue')
        prefetched_findings = prefetched_findings.prefetch_related('test__test_type')
        prefetched_findings = prefetched_findings.prefetch_related('test__engagement__product__jira_pkey_set__conf')
        prefetched_findings = prefetched_findings.prefetch_related('found_by')
        prefetched_findings = prefetched_findings.prefetch_related('risk_acceptance_set')
        # we could try to prefetch only the latest note with SubQuery and OuterRef, but I'm getting that MySql doesn't support limits in subqueries.
        prefetched_findings = prefetched_findings.prefetch_related('notes')
        prefetched_findings = prefetched_findings.prefetch_related('tagged_items__tag')
        prefetched_findings = prefetched_findings.prefetch_related('endpoints')
        prefetched_findings = prefetched_findings.prefetch_related('test__engagement__product__authorized_users')
```

Reply:
Most of my stuff are "inactive", so it would not go through all of them, just a couple thousands. It takes about 5 minutes then on a 5yr old workstation.

Reply:
prefetching didn't work. Do you think it's a blocker for this PR? We could always look afterwards. I've exhausted my time on this PR :-(